hw/drivers/trng/trng_k64f/src/trng_k64f.c

```diff
@@ -0,0 +1,175 @@
+static uint16_t rng_cache_out;
+static uint16_t rng_cache_in;
+static os_stack_t *pstack;
+static bool running;
+
+#define RNGA_POLLER_PRIO (8)
+#define RNGA_POLLER_STACK_SIZE OS_STACK_ALIGN(64)
+static struct os_task poller_task;
+
+static inline void
+k64f_rnga_start(void)
+{
+    RNGA_SetMode(RNG, kRNGA_ModeNormal);
+    running = true;
+}
+
+static inline void
+k64f_rnga_stop(void)
+{
+   RNGA_SetMode(RNG, kRNGA_ModeSleep);
+   running = false;
+}
+
+static size_t
+k64f_trng_read(struct trng_dev *trng, void *ptr, size_t size)
+{
+    os_sr_t sr;
+    size_t num_read;
+
+    OS_ENTER_CRITICAL(sr);
+
+    if (rng_cache_out <= rng_cache_in) {
+        size = min(size, rng_cache_in - rng_cache_out);
+        memcpy(ptr, &rng_cache[rng_cache_out], size);
+        num_read = size;
+    } else if (rng_cache_out + size <= sizeof(rng_cache)) {
+        memcpy(ptr, &rng_cache[rng_cache_out], size);
+        num_read = size;
+    } else {
+        num_read = sizeof(rng_cache) - rng_cache_out;
+        memcpy(ptr, &rng_cache[rng_cache_out], num_read);
+
+        size -= num_read;
+        ptr += num_read;
+
+        size = min(size, rng_cache_in);
+        memcpy(ptr, rng_cache, size);
+        num_read += size;
+    }
+
+    rng_cache_out = (rng_cache_out + num_read) % sizeof(rng_cache);
+
+    if (num_read > 0) {
+        k64f_rnga_start();
+    }
+
+    OS_EXIT_CRITICAL(sr);
+
+    return num_read;
+}
+
+static uint32_t
+k64f_trng_get_u32(struct trng_dev *trng)
+{
+    union {
+        uint32_t v32;
+        uint8_t v8[4];
+    } val;
+    size_t num;
+
+    num = k64f_trng_read(trng, &val.v8, sizeof(val.v8));
+    while (num < sizeof(val.v8)) {
+        os_sched(NULL);
+        num += k64f_trng_read(trng, &val.v8[num], sizeof(val.v8) - num);
+    }
+
+    return val.v32;
+}
+
+static void
+rnga_poller_handler(void *arg)
+{
+    int8_t i;
+    uint8_t data[4];
+    int rc;
+
+    while (1) {
+        if (running) {
+            rc = RNGA_GetRandomData(RNG, data, sizeof(uint32_t));
+            if (rc == 0) {
+                for (i = 0; i < 4; i++) {
+                    rng_cache[rng_cache_in++] = data[i];
+
+                    if (rng_cache_in >= sizeof(rng_cache)) {
+                        rng_cache_in = 0;
+                    }
+
+                    if ((rng_cache_in + 1) % sizeof(rng_cache) == rng_cache_out) {
+                        k64f_rnga_stop();
+                        break;
+                    }
+                }
+            }
+        }
+        os_time_delay(1);
+    }
+}
+
+static int
+k64f_trng_dev_open(struct os_dev *dev, uint32_t wait, void *arg)
+{
+    struct trng_dev *trng;
+
+    trng = (struct trng_dev *)dev;
+    assert(trng);
+
+    if (!(dev->od_flags & OS_DEV_F_STATUS_OPEN)) {
+        rng_cache_out = 0;
+        rng_cache_in = 0;
+
+        RNGA_Init(RNG);
+
+        k64f_rnga_start();
+
+        os_task_init(&poller_task, "rnga_poller", rnga_poller_handler, NULL,
```

Comment:
os_dev_close could be called on dev at some point, then flags could have OS_DEV_F_STATUS_OPEN cleared but task would be already running on second call os_dev_open; then os_task_init could be called second time and that may not be a good idea.

Reply:
I moved the task initialization to _init instead of _open.